pkg/url/url.go

```diff
@@ -396,6 +397,27 @@ func GetValidPortNumber(componentName string, portNumber int, portList []string)
 	return portNumber, fmt.Errorf("given port %d is not exposed on given component, available ports are: %s", portNumber, strings.Trim(strings.Replace(fmt.Sprint(componentPorts), " ", ",", -1), "[]"))
 }
 
+// GetValidExposedPortNumber checks if the given exposed port number is a valid port or not
+// if exposed port is not provided, a random 5 digit valid port will be generated and returned
+func GetValidExposedPortNumber(exposedPort int) (int, error) {
+	// exposed port number will be -1 if the user doesn't specify any port
+	if exposedPort == -1 {
+		freePort, err := util.HttpGetFreePort()
+		if err != nil {
+			return -1, err
+		}
+		return freePort, nil
+	} else {
+		// check if the given port is available
+		listener, err := net.Listen("tcp", ":"+strconv.Itoa(exposedPort))
+		if err != nil {
+			return -1, errors.Wrapf(err, "given port %d is not available, please choose another port", exposedPort)
+		}
+		listener.Close()
```

Comment:
if error is thrown, it won't be reached and thus the listener won't be closed. Thus https://github.com/openshift/odo/pull/2906/files#r412672504